### ISSUE TITLE
Patch for coauthors_plus->fix_author_page() _doing_it_wrong() error

### ISSUE THREAD

**Issue by onpubcom** [Automattic/Co-Authors-Plus#354]
Co-Authors-Plus currently generates the following PHP Notice in the case where the $wp_query global is not yet defined:

`
PHP Notice:  is_author was called <strong>incorrectly</strong>. Conditional query tags do not work before the query is run. Before then, they always return false. Please see <a href="https://codex.wordpress.org/Debugging_in_WordPress">Debugging in WordPress</a> for more information. (This message was added in version 3.1.) in wp-includes/functions.php on line 3897,
PHP Stack trace:,
PHP   1. {main}() index.php:0,
PHP   2. require() index.php:17,
PHP   3. require_once() wp-blog-header.php:13,
PHP   4. require_once() wp-load.php:42,
PHP   5. require_once() wp-config.php:108,
PHP   6. do_action($tag = *uninitialized*, $arg = *uninitialized*) wp-settings.php:277,
PHP   7. call_user_func_array:{wp-includes/plugin.php:525}(*uninitialized*, *uninitialized*) wp-includes/plugin.php:525,
PHP   8. cycloneslider_init(*uninitialized*) wp-includes/plugin.php:525,
PHP   9. CycloneSlider_Plugin->run() plugins/cyclone-slider-2/cyclone-slider.php:160,
PHP  10. CycloneSlider_Plugin->offsetGet($offset = *uninitialized*) plugins/cyclone-slider-2/cyclone-slider.php:35,
PHP  11. cycloneslider_service_export_page($plugin = *uninitialized*) plugins/cyclone-slider-2/src/CycloneSlider/Plugin.php:24,
PHP  12. CycloneSlider_Data->get_sliders($args = *uninitialized*) plugins/cyclone-slider-2/cyclone-slider.php:369,
PHP  13. get_posts($args = *uninitialized*) plugins/cyclone-slider-2/src/CycloneSlider/Data.php:263,
PHP  14. WP_Query->query($query = *uninitialized*) wp-includes/post.php:1763,
PHP  15. WP_Query->get_posts() wp-includes/query.php:4006,
PHP  16. do_action($tag = *uninitialized*, $arg = *uninitialized*) wp-includes/query.php:3409,
PHP  17. call_user_func_array:{wp-includes/plugin.php:525}(*uninitialized*, *uninitialized*) wp-includes/plugin.php:525,
PHP  18. coauthors_plus->fix_author_page(*uninitialized*) wp-includes/plugin.php:525,
PHP  19. is_author($author = *uninitialized*) plugins/co-authors-plus/co-authors-plus.php:937,
PHP  20. _doing_it_wrong($function = *uninitialized*, $message = *uninitialized*, $version = *uninitialized*) wp-includes/query.php:215,
PHP  21. trigger_error(*uninitialized*) wp-includes/functions.php:3897,
`

I'm about to attach a patch that fixes this issue.

**Comment by onpubcom** [comment]
Patch pasted below, sorry too lazy to submit pull request rn :-)

```
Index: co-authors-plus.php
===================================================================
--- co-authors-plus.php	(revision 16716)
+++ co-authors-plus.php	(working copy)
@@ -933,8 +933,9 @@
 	 * Also, we have to do some hacky WP_Query modification for guest authors
 	 */
 	public function fix_author_page() {
+		global $wp_query, $authordata;
 
-		if ( ! is_author() ) {
+		if ( ! isset( $wp_query ) || ! is_author() ) {
 			return;
 		}
 
@@ -945,8 +946,6 @@
 
 		$author = $this->get_coauthor_by( 'user_nicename', $author_name );
 
-		global $wp_query, $authordata;
-
 		if ( is_object( $author ) ) {
 			$authordata = $author;
 			$term = $this->get_author_term( $authordata );
```

**Comment by mjangda** [comment]
Thanks @onpubcom! Can you provide reproduction steps so we can test this out? `fix_author_page` is called in contexts where the query should be set so I'm curious where the issue is.

**Comment by onpubcom** [comment]
Hi @mjangda, this should be reproducible by installing the Cyclone Slider 2 plugin alongside with Co-Authors Plus. Loading any frontend home page will probably trigger this bug with both plugins activated. You can also see some details of what's going on here by looking the xdebug stack trace I posted above.

fix_author_page() is making the assumption that $wp_query has already been setup, since it hooks on to the posts_selection action, which is running before the $wp_query global is set (at least in this particular case).

**Comment by trepmal** [comment]
For convenience, here's my minimized plugin (mu or standard) for reproducing the notice:

```
add_action( 'plugins_loaded', function() {
	get_posts( array(
		'post_type' => 'post',
	) );
} );
```

To be honest, I'm not sure if this would technically be the other plugin's fault for querying so early (I didn't find a definitive guideline on this), but even if so it seems to be an easy win to improve compatibility.  Pull request incoming.

**Comment by onpubcom** [comment]
Thanks, @trepmal

**Comment by douglas-johnson** [comment]
What's up with this issue? I'm getting this warning because of a query made super early by Woocommerce Memberships.

The patch laid out in this issue fixes it for me.

This issue is closed, but pull request #383 that addresses it was never merged.

**Comment by anthonythorne** [comment]
Ive got the same issue on an Admin page, saving an ACF group.

**Comment by GaryJones** [comment]
Connecting the dots - this was addressed in #801.
